README.md

```diff
@@ -1,2 +1,79 @@
-# backoff
-Exponential backoff mechanism
+(API reference)[TODO]
+
+# backoff - exponential backoff mechanism
+
+**backoff** provides an exponential backoff mechanism [1]. It enables
+to reduce contention by making a domain backs off after failing an
+operation contested by another domain, like acquiring a lock or
+performing a `CAS` operation.
+
+The more a domain consecutively failed, the longer it will back off as
+it means the operation is highly contonded. Backing off here means
+calling `Domain.cpu_relax` (or `Thread.yield` if running with a
+version of OCaml previous to 5.0.0).
+
+This mechanism has some drawbacks. First, it adds some delays: for
+example, when a domain releases a contended lock, another domain that
+has backed off after failing acquiring it, will still have to finish
+is back off loop before retrying. Second, this increases any
+unfairness: any other threads that arrives at that time or that has
+failed acquiring the lock for a lesser number of times is more likely
+to acquire it as it will wait less.
+
+## Example
+
+To illustrate how to use `backoff`, here is a small implementation of
+ `Test-and-set` spin lock [2].
+
+```ocaml
+  type t = bool Atomic.t
+
+  let create () = Atomic.make false
+
+  exception Lock_acquired
+
+  let acquire t =
+    let backoff = ref @@ Backoff.create () in
+    try
+      while true do
+        while Atomic.get t do
+	  (* The lock is not available: wait a bit,
+	  before retrying *)
+          backoff := Backoff.once !backoff
+        done;
+	(* Try to acquire the lock, if fail, retry. *)
+        if Atomic.compare_and_set t false true then
+	   raise Lock_acquired
+      done
+    with Lock_acquired -> ()
```

Comment:
I would recommend rewriting this `acquire` example as a recursive function, because it will be more concise and the control flow will also be simpler.

The algorithm can also be improved.  Let me show how I'd roughly write this:

```ocaml
let rec acquire ?(backoff = Backoff.detault) t =
  if Atomic.get t then begin
    Domain.cpu_relax ();
    acquire ~backoff t
  end
  else if not (Atomic.compare_and_set t false true) then
    acquire ~backoff:(Backoff.once backoff) t
```

First of all, in case the first test fails, we only perform a single `cpu_relax ()` call before retrying.  There is no need to backoff more at this point, because we specifically perform only a read-only test of the lock.  We can keep doing this safely in the sense that we are not increasing contention.

Now, in case we see that the lock is released and try to perform a `compare_and_set` then either it will succeed or it will fail.  Failure indicates contention &mdash; there are other domains trying to acquire the lock.  We don't know how many there are, so we'd rather backoff to reduce the chance of having the same problem on next round.

See also [this set of slides](https://cs.brown.edu/courses/cs176/lectures/chapter_07.pdf) that discusses this very topic.